### PR TITLE
Abstraction dev tools

### LINKED ISSUE #12368: Abstraction of bevy_dev_tools
## What problem does this solve or what need does it fill?

Allow for configuration of dev tools

## What solution would you like?

Ideally, we're looking for something that:
- Provides an easy way for user to enable/disable tool
- Is configurable during runtime

## Additional context

see https://github.com/bevyengine/bevy/pull/12354#discussion_r1516637590

### PR DESCRIPTION
**Pull request by matiqo15**
# Objective

- Fixes #12368

## Solution

- Create `DevTool` that holds unique id and state.
- Create `DevToolsStore` resource that stores new `DevTool`s.
- Make existing tool use new abstraction

### CODE REVIEW

File: crates/bevy_dev_tools/src/fps_overlay.rs [hunk 1]
```diff
@@ -26,19 +29,31 @@ pub struct FpsOverlayPlugin {
...
         }
         app.insert_resource(self.config.clone())
+            .register_dev_tool(DevTool::new(Self::FPS_OVERLAY_ID))
             .add_systems(Startup, setup)
             .add_systems(
                 Update,
                 (
                     customize_text.run_if(resource_changed::<FpsOverlayConfig>),
-                    update_text,
+                    change_visibility.run_if(resource_changed::<DevToolsStore>),
```
Comment (matiqo15) [inline crates/bevy_dev_tools/src/fps_overlay.rs:51]: I don't like this. This will run every time any tool has been changed, but I can't think of a solution to only run if a specific tool has been changed

File: crates/bevy_dev_tools/src/lib.rs [hunk 2]
```diff
@@ -45,3 +48,110 @@ impl Plugin for DevToolsPlugin {
         }
     }
 }
+
+/// Unique identifier for [`DevTool`].
+#[derive(Debug, Hash, Eq, PartialEq, Clone)]
+pub struct DevToolId(pub Uuid);
```
Comment (pablo-lua) [inline crates/bevy_dev_tools/src/lib.rs:54]: I'm wondering: Why not use a TypeId map instead of Uuid? I think its much more simple.

File: crates/bevy_dev_tools/src/lib.rs [hunk 3]
```diff
@@ -45,3 +48,110 @@ impl Plugin for DevToolsPlugin {
...
+/// Extend [`App`] with new `register_dev_tool` function.
+pub trait RegisterDevTool {
+    /// Registers a new [`DevTool`].
+    fn register_dev_tool(&mut self, dev_tool: DevTool) -> &mut Self;
+}
+
+impl RegisterDevTool for App {
+    fn register_dev_tool(&mut self, dev_tool: DevTool) -> &mut Self {
+        self.init_resource::<DevToolsStore>();
+        let mut dev_tools = self.world.resource_mut::<DevToolsStore>();
```
Comment (pablo-lua) [inline crates/bevy_dev_tools/src/lib.rs:153]:
```suggestion
        let mut dev_tools = self.world.get_resource_or_insert_with::<DevToolsStore>(Default::default);
```

Comment (pablo-lua) [reply crates/bevy_dev_tools/src/lib.rs:153]: Note: This can cause Ci formatter error, not sure.

File: crates/bevy_dev_tools/src/fps_overlay.rs [hunk 4]
```diff
@@ -96,3 +111,21 @@ fn customize_text(
         }
     }
 }
+
+fn change_visibility(
+    mut query: Query<&mut Visibility, With<FpsText>>,
```
Comment (pablo-lua) [inline crates/bevy_dev_tools/src/fps_overlay.rs:116]: Can't this cause conflict error everywhere when the user uses mut Visibility? Or will not cause conflict because Query know when archetypes cross each other?

File: crates/bevy_dev_tools/src/fps_overlay.rs (hunk 1 above)
Comment (pablo-lua) [reply crates/bevy_dev_tools/src/fps_overlay.rs:51]: I think we can relax on that somehow by just making a global system that manages this, based on one-shot systems, I'm not sure how effective this can be do though. At the same time, that would be useless without some kind o cache to store what was changed and what was not.

Comment (pablo-lua) [reply crates/bevy_dev_tools/src/fps_overlay.rs:51]: I think we can relax on that somehow by just making a global system that manages this, based on one-shot systems, I'm not sure how effective this can be do though. At the same time, that would be useless without some kind o cache to store what was changed and what was not.

Comment (matiqo15) [reply crates/bevy_dev_tools/src/fps_overlay.rs:51]: If we had some kind of cache storing what was changed, I don't think there would be any reason for complicating it with the global system. I'd just let the user interact with the cache and use it in a `run_if` condition.

File: crates/bevy_dev_tools/src/fps_overlay.rs (hunk 4 above)
Comment (matiqo15) [reply crates/bevy_dev_tools/src/fps_overlay.rs:116]: From what I know this shouldn't cause any issues.

File: crates/bevy_dev_tools/src/lib.rs (hunk 2 above)
Comment (matiqo15) [reply crates/bevy_dev_tools/src/lib.rs:54]: Do we want to identify each dev tool with a different type? We could use plugin struct for it, but what if plugin adds more than one dev tool?

Comment (pablo-lua) [reply crates/bevy_dev_tools/src/lib.rs:54]: I was expecting to identify every tool with a single config type registered to it, but I see that's a similar approach taken by the gizmos multiple configs, and we aren't storing configs in this map by now, so, with we have now on this PR, the TypeId doesn't make sense.
I can see number of reasons why offering a config type though. For example, we might want to do more than enable and disable things, we might want this DevToolsStore to store entire configurations in the future, like storing the FpsOverlayConfig here, and let the user refer to it from the TypeId (which can be easily created and so).

File: crates/bevy_dev_tools/src/fps_overlay.rs (hunk 4 above)
Comment (alice-i-cecile) [reply crates/bevy_dev_tools/src/fps_overlay.rs:116]: Yeah, this will generally run in parallel. Those checks are based on "factual compatibility": they check the list of actual archetypes.

File: crates/bevy_dev_tools/src/fps_overlay.rs (hunk 1 above)
Comment (alice-i-cecile) [reply crates/bevy_dev_tools/src/fps_overlay.rs:51]: I think this is a fine level of granularity for now: I wouldn't expect `DevToolsStore` to change regularly at all.

File: crates/bevy_dev_tools/src/lib.rs [hunk 5]
```diff
@@ -45,3 +48,111 @@ impl Plugin for DevToolsPlugin {
...
+    pub fn add(&mut self, dev_tool: DevTool) {
+        self.dev_tools.insert(dev_tool.id.clone(), dev_tool);
+    }
+
+    /// Removes a [`DevTool`].
+    pub fn remove(&mut self, id: &DevToolId) {
+        self.dev_tools.remove(id);
+    }
+
+    /// Returns a [`DevTool`].
```
Comment (alice-i-cecile) [inline crates/bevy_dev_tools/src/lib.rs:123]:
```suggestion
    /// Returns a reference to the given [`DevTool`] if present.
```

File: crates/bevy_dev_tools/src/lib.rs [hunk 6]
```diff
@@ -45,3 +48,111 @@ impl Plugin for DevToolsPlugin {
...
+    pub fn get(&self, id: &DevToolId) -> Option<&DevTool> {
+        self.dev_tools.get(id)
+    }
+
+    /// Returns a mutable [`DevTool`].
```
Comment (alice-i-cecile) [inline crates/bevy_dev_tools/src/lib.rs:128]:
```suggestion
    /// Returns a mutable reference to the given [`DevTool`] if present.
```

File: crates/bevy_dev_tools/src/lib.rs [hunk 7]
```diff
@@ -45,3 +48,111 @@ impl Plugin for DevToolsPlugin {
...
+    pub fn get_mut(&mut self, id: &DevToolId) -> Option<&mut DevTool> {
+        self.dev_tools.get_mut(id)
+    }
+
+    /// Returns an iterator over all [`DevTool`]s.
```
Comment (alice-i-cecile) [inline crates/bevy_dev_tools/src/lib.rs:133]:
```suggestion
    /// Returns an iterator over all [`DevTool`]s, by reference.
```

**Review by alice-i-cecile — CHANGES_REQUESTED** [review]
I finally have the energy to review this: sorry for keeping you waiting! Generally impressed with the polish and thought here, but I think we should try harder to remove the need for UUIDs and try to be more consistent with other parts of the engine.

I like:

- .register_dev_tool
- a central registry
- a way to iterate over all of the devtools
- pulling shared config like "is_enabled" out into the struct

I don't like:
- the need to use UUIDs
- the fact that it's unclear where we should store tool-specific config

I think we should:

- borrow from the [`Asset`] trait heavily, and create a matching `DevTool` trait
- use the type that implements `DevTool` to store tool-specific config
- add a `DevTool: Reflect` bound: this will be essential for editor use cases and we need to make sure it works *now*
- use TypeId keys

In summary:

1. Every dev tool has its own config type, which may be a unit struct.
2. The `DevTool` trait is implemented for these types.
3. `DevToolConfig` stores both common settings (e.g. is_enabled) and tool specific settings, using a `Box<dyn DevTool>`.
4. `DevToolsStore` is a wrapper over `HashMap<TypeId, DevToolConfig>`.
5. We add dev tools using `app.init_dev_tool::<D: DevTool>`, with a matching `app.insert_dev_tool(d: D)` method for when you want to supply config.

File: crates/bevy_dev_tools/src/lib.rs [hunk 8]
```diff
@@ -45,3 +49,128 @@ impl Plugin for DevToolsPlugin {
...
+        let dev_tool = DevToolConfig::new(TypeId::of::<D>(), D::default());
+        let mut dev_tools = self
+            .world
+            .get_resource_or_insert_with::<DevToolsStore>(Default::default);
+        dev_tools.add(dev_tool);
+        self
+    }
+
+    fn insert_dev_tool<D: DevTool>(&mut self, value: D) -> &mut Self {
+        let dev_tool = DevToolConfig::new(TypeId::of::<D>(), value);
```
Comment (pablo-lua) [inline crates/bevy_dev_tools/src/lib.rs:169]:
```suggestion
    fn insert_dev_tool<D: DevTool>(&mut self, tool_config: D) -> &mut Self {
        let dev_tool = DevToolConfig::new(TypeId::of::<D>(), tool_config);
```

Comment (pablo-lua) [reply crates/bevy_dev_tools/src/lib.rs:169]: Just a better naming here, but can be a opinion question.

File: crates/bevy_dev_tools/src/lib.rs [hunk 9]
```diff
@@ -45,3 +49,128 @@ impl Plugin for DevToolsPlugin {
...
+        self.is_enabled = false;
+    }
+
+    /// Toggles [`DevTool`].
+    pub fn toggle(&mut self) {
+        self.is_enabled = !self.is_enabled;
+    }
+}
+
+impl DevToolConfig {
```
[the remark this answers is not in the document]
Comment (pablo-lua) [reply crates/bevy_dev_tools/src/lib.rs:88]: Also, I think we might want helper methods to return the inner config as &dyn Reflect and &mut dyn Reflect.

**Review by pablo-lua — CHANGES_REQUESTED** [review]
Looks good! Just a few comments and a request: Can you change the new ui debug overlay to use this abstraction? see #10420

**Comment by cart** [comment]
I'm not convinced we should have a "dev tool" abstraction like this:

1. What constitutes a "dev tool" is extremely nebulous. I do not think there is a good clear line between a "dev tool" and a "bevy engine feature".
2. This moves configuration outside of the standard "ECS resource" pattern and into a consolidated DevToolStore resource. This increases the clunkiness of accessing dev tool config significantly and makes it impossible to use ECS change detection to determine when the config has changed.
3. If we want to expose an existing Bevy feature in some "dev tool debug interface", we shouldn't need to refactor how it is configured.
4. The motivation appears to be "we want a queryable catalog of dev tools and their configuration, with a common on/off switch". I would argue that "Dev tool" is too general of a concept to be a useful generic "category of thing". Instead, consumers of Bevy features (such as a "dev tool debug interface") should be responsible for choosing what Bevy features to expose and how to expose them (including what it means to enable and disable a feature). The "dev tool debug interface" can be responsible for being the catalog of features to expose.
5. If we really need a "common on/off switch", that is something that should be considered holistically. However without consideration for the lifecycle of adding/removing features (systems, config, etc), I don't think this would provide significant value. Instead, just add `enabled: bool` to whatever feature-config resource makes sense. It should be easy to bind specific bool values to whatever debug interface we are building.

I think, if we were to embrace a generic pattern like this, it should be done at the Plugin level and after significant consideration about "plugin/feature configuration and lifecycles".

**Comment by alice-i-cecile** [comment]
@cart I understand your concerns. I'll aim to write up a proper response and design vision in an RFC, and link it here (and the linked issue). I think there's very good reasons to want a cohesive method of control (even if this might not be the best possible design), but those aren't clearly articulated anywhere.

**Comment by NthTensor** [comment]
Closing, as the dev-tools abstraction effort petered out. This may be resurrected as part of a future project goal.
